### PR TITLE
Chef extension support for existing Azure ASM deployment.

### PR DESCRIPTION
**Pull request by Aliasgar16**
In-progress.

### CODE REVIEW

**Comment by Aliasgar16** [comment]
Changes along with test is still in-progress.

**Comment by Aliasgar16** [comment]
Testing is in-progress.

**Comment by Aliasgar16** [comment]
Resolved bugs in the code and handled some corner cases. Successfully tested the functionality on Linux and Windows platform. RSpecs are in-progress.

**Comment by Aliasgar16** [comment]
RSpecs for `setup_extension`, `update_role_xml` and role `update` methods are pending.

**Comment by Aliasgar16** [comment]
Completed with all the RSpecs. Testing done. Overall: Done.

File: lib/azure/service_management/role.rb [hunk 1]
```diff
@@ -548,5 +562,144 @@ def create(params, roleXML)
...
+            xml.OSVirtualHardDisk role_xml.at_css('OSVirtualHardDisk').children if !role_xml.at_css('OSVirtualHardDisk').nil?
+            xml.RoleSize role_xml.at_css('RoleSize').children if !role_xml.at_css('RoleSize').nil?
+            xml.ProvisionGuestAgent role_xml.at_css('ProvisionGuestAgent').children if !role_xml.at_css('ProvisionGuestAgent').nil?
+        }
+      end
+
+      builder.doc.to_xml.gsub("&lt\;","<").gsub("&gt\;",">")
+    end
+
+    def update_role_xml(roleXML, params)
```
Comment (NimishaS) [inline lib/azure/service_management/role.rb]: please change this name to `update_role_xml_for_extension`

File: lib/azure/service_management/role.rb [hunk 2]
```diff
@@ -548,5 +562,144 @@ def create(params, roleXML)
       "/#{params['deploy_name']}/roles"
       @connection.query_azure(servicecall, "post", roleXML.to_xml)
     end
+
+    def setup_extension(params)
+      role_xml = update_role_xml(params[:role_xml], params) ## add Chef Extension config in VM's role_xml
+
+      ## create new XML (with Chef Extension config and other pre-existing VM config) using the role_xml retrieved from the VM
```
Comment (NimishaS) [inline lib/azure/service_management/role.rb]: Please add comment specifying why we can't use the `role_xml` retrieved above as it is and have to create a New XML

**Comment by Aliasgar16** [comment]
Fixed review comments and did testing.

**Comment by adamedx** [comment]
So I tried to run this, but hit an error pretty early on with all knife commands:

```
adamed@bantu:~/src/knife-azure$ bundle exec knife bootstrap --help
/home/adamed/src/knife-azure/lib/chef/knife/azurerm_server_delete.rb:31:in `<class:AzurermServerDelete>': uninitialized constant Chef::Knife::AzurermBase (NameError)
        from /home/adamed/src/knife-azure/lib/chef/knife/azurerm_server_delete.rb:29:in `<class:Knife>'
        from /home/adamed/src/knife-azure/lib/chef/knife/azurerm_server_delete.rb:28:in `<class:Chef>'
        from /home/adamed/src/knife-azure/lib/chef/knife/azurerm_server_delete.rb:27:in `<top (required)>'
        from /home/adamed/.chefdk/gem/ruby/2.1.0/gems/chef-12.9.38/lib/chef/knife/core/subcommand_loader.rb:100:in `load'
        from /home/adamed/.chefdk/gem/ruby/2.1.0/gems/chef-12.9.38/lib/chef/knife/core/subcommand_loader.rb:100:in `block in load_commands'
        from /home/adamed/.chefdk/gem/ruby/2.1.0/gems/chef-12.9.38/lib/chef/knife/core/subcommand_loader.rb:100:in `each'
        from /home/adamed/.chefdk/gem/ruby/2.1.0/gems/chef-12.9.38/lib/chef/knife/core/subcommand_loader.rb:100:in `load_commands'
        from /home/adamed/.chefdk/gem/ruby/2.1.0/gems/chef-12.9.38/lib/chef/knife/core/subcommand_loader.rb:110:in `load_command'
        from /home/adamed/.chefdk/gem/ruby/2.1.0/gems/chef-12.9.38/lib/chef/knife/core/subcommand_loader.rb:124:in `command_class_from'
        from /home/adamed/.chefdk/gem/ruby/2.1.0/gems/chef-12.9.38/lib/chef/knife.rb:153:in `subcommand_class_from'
        from /home/adamed/.chefdk/gem/ruby/2.1.0/gems/chef-12.9.38/lib/chef/knife.rb:214:in `run'
        from /home/adamed/.chefdk/gem/ruby/2.1.0/gems/chef-12.9.38/lib/chef/application/knife.rb:148:in `run'
        from /home/adamed/.chefdk/gem/ruby/2.1.0/gems/chef-12.9.38/bin/knife:25:in `<top (required)>'
        from /home/adamed/.chefdk/gem/ruby/2.1.0/bin/knife:22:in `load'
        from /home/adamed/.chefdk/gem/ruby/2.1.0/bin/knife:22:in `<main>'
adamed@bantu:~/src/knife-azure$ bundle exec knife bootstrap --help```

**Comment by Aliasgar16** [comment]
@adamedx I am not hitting any such error on this branch `ali/add_extension_to_existing_vm`. Which branch you are on ?.  May be the branch from which you are trying the command is not updated or has some issues or conflicts. Also the command implemented in this PR is `knife bootstrap azure` and not Chef's standard command `knife bootstrap`.

**Comment by adamedx** [comment]
Still repros for me after narrowing things down and using only released software. Steps on Ubuntu 15:

1. Install latest chef-dk (0.12.0)
2. Add chef-dk's ruby to your path with `chef shell-init bash`
3. Check out this branch
4. Remove any gemfile.lock
5. `bundle install`
6. Run `bundle exec knife bootstrap --help`

Here's some of the output, minus the gemfile.lock and bundle install stuff:

```
adamed@bantu:~$ eval "$(/opt/chefdk/bin/chef shell-init bash)"                       
adamed@bantu:~$ chef -v                                                              
Chef Development Kit Version: 0.12.0                                                 
chef-client version: 12.8.1                                                          
berks version: 4.3.0                                                                 
kitchen version: 1.6.0                                                               
adamed@bantu:~$ cd ~/src/knife-azure                                                 
adamed@bantu:~/src/knife-azure$ git status                                           
On branch ali/add_extension_to_existing_vm                                           
Your branch is up-to-date with 'origin/ali/add_extension_to_existing_vm'.            
nothing to commit, working directory clean                                           
adamed@bantu:~/src/knife-azure$ git log -n 1                                         
commit 3f180efea274cb7d643835159a797fd9585db7e6                                      
Author: aliasgar16 <aliasgar.batterywala@clogeny.com>                                
Date:   Tue Apr 12 17:45:32 2016 +0530                                               
                                                                                     
    Fixed review comments and did testing.                                           
adamed@bantu:~/src/knife-azure$ bundle exec knife bootstrap --help                   
/home/adamed/src/knife-azure/lib/chef/knife/azurerm_server_delete.rb:31:in `<class:Az
urermServerDelete>': uninitialized constant Chef::Knife::AzurermBase (NameError)     
        from /home/adamed/src/knife-azure/lib/chef/knife/azurerm_server_delete.rb:29:
in `<class:Knife>'                                                                   
        from /home/adamed/src/knife-azure/lib/chef/knife/azurerm_server_delete.rb:28:
in `<class:Chef>'                                                                    
        from /home/adamed/src/knife-azure/lib/chef/knife/azurerm_server_delete.rb:27:
in `<top (required)>'                                                                ```

**Comment by Aliasgar16** [comment]
@adamedx I tried to reproduce the issue using your steps, but I could not reproduce it. Instead during the `bundle install` command the installation for gem named `listen (3.1.1)` failed as it required `Ruby 2.2`, while `chef-dk (0.12.0)` has embedded `Ruby 2.1`, so I used older version of this gem (`listen 3.0.6`, which did not had this `Ruby 2.2` dependency), doing this caused `bundle install` to succeed and also other `knife commands` worked. Could you try to clean up and re-configure things at your end.

**Comment by adamedx** [comment]
Let's rebase first @Aliasgar16 and I'll try again.

**Comment by Aliasgar16** [comment]
@adamedx I have `re-based` this branch. Also, I tried to reproduce this issue on a `separate Azure VM (Ubuntu 15.10)` where I installed and configured the required things like `ChefDK (0.12.0)` from scratch and used `ChefDK's Ruby 2.1`, but unfortunately I could not reproduce it.

**Comment by adamedx** [comment]
Ok, the errors are fixed, but I still can't run the command -- I get the following error:

```bundle exec knife bootstrap azure adamedlxk -r 'devbox'
ERROR: You did not provide a valid 'Azure Dns Name' value. Please set knife[:azure_dns_name] in your knife.rb or pass as an option.```

**Comment by Aliasgar16** [comment]
@adamedx `azure_dns_name` is a compulsory option to run this command. Please see the below sample command for your reference:

`bundle exec knife bootstrap azure ali-ubu-test --azure-dns-name ali-ubu-test -VV`

**Comment by adamedx** [comment]
So to be clear,  Azure xplat-cli does not require you to specify the dns name when adding the extension, nor does the powershell cmdlet, and in general we don't require you to do this with `knife azure server delete`.

We should do whatever those commands are doing to avoid passing this extra parameter.

**Comment by Aliasgar16** [comment]
@adamedx I have made the necessary changes to keep `azure_dns_name` as an optional parameter. Please let me know your feedback.

**Comment by adamedx** [comment]
All right, it works for me , so 👍 . Both of the following worked:

`knife bootstrap azure myexistingvm`

and

`knife bootstrap azure myexistingvm -r 'myrunlist'`

Note that this is much easier than bootstrapping via `ssh` or `winrm` because you don't need to specify credentials or know the publicly accessible ip (there may not even be one). 

So nice work. :)
